core/src/consensus.rs

```diff
@@ -642,6 +642,22 @@ impl Tower {
                     SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor)
                 };
 
+                // `heaviest_subtree_fork_choice` entries are not cleaned by duplicate block purging/rollback logic,
+                // so this is safe to check here. We return here if the last voted slot was rolled back/purged due to
+                // being a duplicate because `ancestors`/`descendants`/`progress` structurs may be missing this slot due
+                // to duplicate purging. This would cause many of the `unwrap()` checks below to fail.
+                if let Some(latest_duplicate_ancestor) = heaviest_subtree_fork_choice.latest_invalid_ancestor(&(last_voted_slot, last_voted_hash)) {
+                    // We're rolling back because one of the ancestors of the last vote was a duplicate. In this
+                    // case, it's acceptable if the switch candidate is one of ancestors of the previous vote,
+                    // just fail the switch check because there's no point in voting on an ancestor. ReplayStage
+                    // should then have a special case continue building an alternate fork from this ancestor, NOT
+                    // the `last_voted_slot`. This is in contrast to usual SwitchFailure where ReplayStage continues to build blocks
+                    // on latest vote. See `select_vote_and_reset_forks()` for more details.
+                    if heaviest_subtree_fork_choice.is_ancestor_slot(switch_slot, &(last_voted_slot, last_voted_hash)) {
+                        return rollback_due_to_to_to_duplicate_ancestor(latest_duplicate_ancestor);
+                    }
```

Comment:
Yeah you're talking about a case like this right?

```
                     0
            /            \
         1                3 (correct dup version that was duplicate confirmed)
      /                       \
   3 (our last vote)  6 (< Switch threshold have voted on this)
 /
5
```

We marked  our fork `1->3`duplicate, dumped it, and repaired `3->6`. We try to switch to slot `6` and this check `switch_slot_ancestors.contains(&last_voted_slot)` here https://github.com/solana-labs/solana/blob/b2d071a9ea01c622b90cbece49f313b5bb886afb/core/src/consensus.rs#L680  returns true, even though this fork did not reach switching threshold.

When thinking about this situation, there are a few cases to consider:

1) Our last vote was on our dumped version of`3`
2) Our last vote was on a descendant of the dumped version of `3`.

We know this new version of  slot `3` must be duplicate confirmed so it has at least 52% of the votes on it, which should be enough to generate a switching proof if your last vote was on your version of `3`. In case 2 however, it's possible your last vote was on `5` in the diagram above, in which case the lockouts on those observed votes for the duplicate threshold do not extend past `5`. In these cases I think we have either have to:

1) Distinguish between the old version of `3` and the new version of `3` using the fork choice (which is why we don't purge the fork choice, it keeps the entire history around).  However, then what happens if we restart and lose the fork choice tree? 
2) Wait for everybody to vote past slot `5` on the new fork at which time we know *for sure* the lockout is past 3. 

In both cases we won't halt waiting because at least 52% of the network has voted on the correct version of 3, so they should be on the right fork


Basically my approach right now is to ignore these cases and allow a "freebie" vote that may violate switching proofs in the case of duplicates. Later we'll need to patch all these edge cases up.

Reply:
I don't think that situation 2) is a concern because it wouldn't pass this check `switch_slot_ancestors.contains(&last_voted_slot)` to return `SameFork`. I was concerned about situation 1) and bypassing lockout, wasn't thinking about switch threshold but that's a good point too. Curious what you think about the lockout, but for the switch threshold it would be good to document the freebie vote issue

Reply:
Yup I made a special case for that freebie now, and added some tests for these edge cases: https://github.com/solana-labs/solana/pull/18160/files#diff-6d8458bb2e53158ac472a9ad4709e6a0a52b75d930c019013298e8acda133828R4844-R5184. 

I found an additional bug in the rollback + lockout scenario illustrated by this ignored test: https://github.com/solana-labs/solana/pull/18160/files#diff-6d8458bb2e53158ac472a9ad4709e6a0a52b75d930c019013298e8acda133828R5166-R5167

The issue stems from the fact that the locked out check can't distinguish between two ancestors with the same slot but different hashes. This `contains` check can pass even though we are on a different fork: https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L538